compiler/rustc_trait_selection/src/traits/specialize/specialization_graph.rs

```diff
@@ -94,128 +95,160 @@ impl<'tcx> ChildrenExt<'tcx> for Children {
-                let ge = tcx.specializes((possible_sibling, impl_def_id));
+                return Ok((false, false));
+            }
 
-                if le == ge { report_overlap_error(overlap, last_lint_mut) } else { Ok((le, ge)) }
-            })?;
+            let le = tcx.specializes((impl_def_id, possible_sibling));
+            let ge = tcx.specializes((possible_sibling, impl_def_id));
 
-            if le && !ge {
-                debug!(
-                    "descending as child of TraitRef {:?}",
-                    tcx.impl_trait_ref(possible_sibling).unwrap().subst_identity()
-                );
+            if le == ge { report_overlap_error(overlap, last_lint_mut) } else { Ok((le, ge)) }
+        })?;
 
-                // The impl specializes `possible_sibling`.
-                return Ok(Inserted::ShouldRecurseOn(possible_sibling));
-            } else if ge && !le {
-                debug!(
-                    "placing as parent of TraitRef {:?}",
-                    tcx.impl_trait_ref(possible_sibling).unwrap().subst_identity()
-                );
+        if le && !ge {
+            debug!(
+                "descending as child of TraitRef {:?}",
+                tcx.impl_trait_ref(possible_sibling).unwrap().subst_identity()
+            );
 
-                replace_children.push(possible_sibling);
-            } else {
-                // Either there's no overlap, or the overlap was already reported by
-                // `overlap_error`.
-            }
-        }
+            // The impl specializes `possible_sibling`.
+            return Ok(OverlapResult::SpecializeOne(possible_sibling));
+        } else if ge && !le {
+            debug!(
+                "placing as parent of TraitRef {:?}",
+                tcx.impl_trait_ref(possible_sibling).unwrap().subst_identity()
+            );
 
-        if !replace_children.is_empty() {
-            return Ok(Inserted::ReplaceChildren(replace_children));
+            replace_children.push(possible_sibling);
+        } else {
+            // Either there's no overlap, or the overlap was already reported by
+            // `overlap_error`.
         }
+    }
 
-        // No overlap with any potential siblings, so add as a new sibling.
-        debug!("placing as new sibling");
-        self.insert_blindly(tcx, impl_def_id);
-        Ok(Inserted::BecameNewSibling(last_lint))
+    if !replace_children.is_empty() {
+        return Ok(OverlapResult::SpecializeAll(replace_children));
     }
+
+    if overlap_mode.use_negative_impl() {
+        if tcx.impl_polarity(impl_def_id) == ty::ImplPolarity::Positive {
+            let trait_ref = tcx.impl_trait_ref(impl_def_id).unwrap().skip_binder();
+
+            if traits::negative_impl_holds(tcx, impl_def_id, overlap_mode) {
+                return Err(OverlapError {
+                    with_impl: impl_def_id,
+                    trait_ref,
+                    self_ty: None,
```

Comment:
You should be able to extract the `self_ty` from the trait ref above. It'll prove the error message from:

> conflicting implementations of trait `MyTrait`

to 

> conflicting implementations of trait `MyTrait` for `Foo`